diff --git a/src/browser.c b/src/browser.c
index f12573a4..bf15e9d1 100644
--- a/src/browser.c
+++ b/src/browser.c
@@ -679,6 +679,9 @@ char *browse_in(const char *inpath)
 	char *path = real_dir_from_tilde(inpath);
 	struct stat fileinfo;
 
+    /* Open the files relative to the open file*/
+    if (strcmp(path, "") == 0) path = free_and_assign(path, copy_of(openfile->filename));
+
 	/* If path is not a directory, try to strip a filename from it; if then
 	 * still not a directory, use the current working directory instead. */
 	if (stat(path, &fileinfo) == -1 || !S_ISDIR(fileinfo.st_mode)) {
