--- a/mods/src/misc/misc.c
+++ b/mods/src/misc/misc.c
@@ -117,6 +117,11 @@ static unsigned char *RakNetInstance_injection(unsigned char *rak_net_instance)
     return result;
 }
 
+// Increase Player Speed
+static float Player_getWalkingSpeedModifier_injection(__attribute__((unused)) unsigned char *player) {
+    return 1.3; // Default Is 1
+}
+
 // Init
 void init_misc() {
     if (feature_has("Remove Invalid Item Background", 0)) {
@@ -144,6 +149,9 @@ void init_misc() {
     // Fix Bug Where RakNetInstance Starts Pinging Potential Servers Before The "Join Game" Screen Is Opened
     overwrite_calls((void *) RakNetInstance, (void *) RakNetInstance_injection);
 
+    // Increase Player Speed
+    overwrite((void *) Player_getWalkingSpeedModifier, (void *) Player_getWalkingSpeedModifier_injection);
+
     // Init C++
     _init_misc_cpp();
 }
